README.rst
roughrider.application
**********************

Base WSGI application using ``horseman`` nodes with routing capabilities.


Example
=======

Below is an example of an application, handling a GET request on '/'
and returning a JSON response.

.. code-block:: python

  from horseman.response import Response
  from roughrider.application import Application

  app = Application()

  @app.routes.register('/')
  def json(request):
      return Response.to_json(body={'message': "Hello, world!"})


docs/HISTORY.rst
CHANGES
=======

0.3.1 (2022-02-22)
------------------

  * Added a handful of tests for `application_uri`.


0.3 (2022-02-22)
----------------

  * Request `application_uri` now handles environs that are not well formed.
    It fixes some uses of "proxy" or "virtual host" middlewares.


0.2 (2022-02-11)
----------------

  * Corrected the request `extract` method.


0.1 (2021-12-27)
----------------

  * Initial release.


tests/test_request.py
import pytest
import webtest.app
from io import BytesIO
from horseman.parsers import Data
from roughrider.application import Application
from roughrider.application.request import Request


def test_request_defaults():
    app = Application()
    environ = webtest.app.TestRequest.blank('/test', method='GET').environ
    request = Request(app, environ)
    assert request.method == 'GET'
    assert request.script_name == ''
    assert request.query == {}
    assert request.cookies == {}
    assert request.content_type == None
    assert request.application_uri == 'http://localhost'
    assert request.uri() == 'http://localhost/test'


def test_request_queryvars():
    app = Application()
    environ = webtest.app.TestRequest.blank(
        '/test?foo=1&bar=2', method='GET').environ
    request = Request(app, environ)
    assert request.query == {'bar': ['2'], 'foo': ['1']}
    assert request.query.to_dict() == {'bar': '2', 'foo': '1'}
    assert request.uri() == 'http://localhost/test?foo%3D1%26bar%3D2'
    assert request.uri(include_query=False) == 'http://localhost/test'


def test_request_uri():
    environ = {
        'REQUEST_METHOD': 'GET',
        'SCRIPT_NAME': '',
        'PATH_INFO': '/',
        'QUERY_STRING': '',
        'SERVER_NAME': 'test_domain.com',
        'SERVER_PORT': '80',
        'HTTP_HOST': 'test_domain.com:80',
        'SERVER_PROTOCOL': 'HTTP/1.0',
        'wsgi.url_scheme': 'http',
        'CONTENT_TYPE': "application/json",
        'QUERY_STRING': 'action=login&token=abcdef',
        'wsgi.input': BytesIO(
            b'''{"username": "test", "password": "test"}'''
        ),
        'wsgi.errors': BytesIO()
    }

    app = Application()
    request = Request(app, environ)
    assert request.uri() == (
        'http://test_domain.com/?action%3Dlogin%26token%3Dabcdef'
    )
    assert request.uri(include_query=False) == 'http://test_domain.com/'

    environ['SCRIPT_NAME'] = '/api'
    request = Request(app, environ)
    assert request.uri() == (
        'http://test_domain.com/api/?action%3Dlogin%26token%3Dabcdef'
    )
    assert request.uri(include_query=False) == 'http://test_domain.com/api/'

    environ['wsgi.url_scheme'] = 'https'
    environ['HTTP_HOST'] = 'example.com:443'
    environ['SERVER_PORT'] = '443'
    request = Request(app, environ)
    assert request.uri() == (
        'https://example.com/api/?action%3Dlogin%26token%3Dabcdef'
    )
    assert request.uri(include_query=False) == 'https://example.com/api/'

    environ['HTTP_HOST'] = 'example.com:999'
    environ['SERVER_PORT'] = '999'
    request = Request(app, environ)
    assert request.uri() == (
        'https://example.com:999/api/?action%3Dlogin%26token%3Dabcdef'
    )
    assert request.uri(include_query=False) == 'https://example.com:999/api/'

    environ['wsgi.url_scheme'] = 'http'
    request = Request(app, environ)
    assert request.uri() == (
        'http://example.com:999/api/?action%3Dlogin%26token%3Dabcdef'
    )
    assert request.uri(include_query=False) == 'http://example.com:999/api/'

    # Faulty environ. Needs to handle things like repoze.vhm
    environ['HTTP_HOST'] = 'example.com'
    del environ['SERVER_PORT']
    request = Request(app, environ)
    assert request.uri() == (
        'http://example.com/api/?action%3Dlogin%26token%3Dabcdef'
    )

    del environ['HTTP_HOST']
    environ['SERVER_NAME'] = 'example.com'
    request = Request(app, environ)
    assert request.uri() == (
        'http://example.com/api/?action%3Dlogin%26token%3Dabcdef'
    )


def test_request_cookies():
    app = Application()
    environ = webtest.app.TestRequest.blank(
        '/test', cookies={'key': 'value'}).environ
    request = Request(app, environ)
    assert request.cookies == {'key': 'value'}


def test_request_data():
    environ = {
        'REQUEST_METHOD': 'GET',
        'SCRIPT_NAME': '',
        'PATH_INFO': '/',
        'QUERY_STRING': '',
        'SERVER_NAME': 'test_domain.com',
        'SERVER_PORT': '80',
        'HTTP_HOST': 'test_domain.com:80',
        'SERVER_PROTOCOL': 'HTTP/1.0',
        'wsgi.url_scheme': 'http',
        'CONTENT_TYPE': "application/json",
        'QUERY_STRING': 'action=login&token=abcdef',
        'wsgi.input': BytesIO(
            b'''{"username": "test", "password": "test"}'''
        ),
        'wsgi.errors': BytesIO()
    }
    app = Application()
    request = Request(app, environ)
    data = request.extract()

    assert data == Data(
        form=None,
        files=None,
        json={'username': 'test', 'password': 'test'}
    )

    assert request.extract() is data


tests/test_application.py
import pytest
from horseman.http import HTTPError
from horseman.response import Response
from roughrider.application import Application
from roughrider.routing.components import NamedRoutes
from roughrider.application.request import Request


def test_application_defaults():
    app = Application()
    assert isinstance(app.routes, NamedRoutes)
    assert app.utilities == {}
    assert app.request_factory is Request

    app = Application(utilities={'emailer': 1})
    assert app.utilities == {'emailer': 1}


def test_application_resolver():
    app = Application()

    @app.routes.register('/test')
    def endpoint(request):
        return Response(200, body=b'Test')

    environ = {
        'REQUEST_METHOD': 'GET',
        'SCRIPT_NAME': '',
        'PATH_INFO': '/',
        'QUERY_STRING': '',
        'SERVER_NAME': 'test_domain.com',
        'SERVER_PORT': '80',
        'HTTP_HOST': 'test_domain.com:80',
        'SERVER_PROTOCOL': 'HTTP/1.0',
        'wsgi.url_scheme': 'http',
    }
    response = app.resolve('/', environ)
    assert response is None

    response = app.resolve('/test', environ)
    assert response is not None
    assert isinstance(response, Response)
    assert response.body == b'Test'

    environ['REQUEST_METHOD'] = 'POST'
    with pytest.raises(HTTPError) as exc:
        app.resolve('/test', environ)

    assert str(exc.value) == 'HTTPStatus.METHOD_NOT_ALLOWED'


tests/test_middlewares.py
import pytest
import logging
import webtest.app
from horseman.response import Response
from roughrider.application import WrappableApplication


def auth_middleware(app):
    def auth_wrapping(environ, start_response):
        if environ.get('REMOTE_USER') == 'admin':
            return app(environ, start_response)
        return Response(401)(environ, start_response)
    return auth_wrapping


def log_middleware(app):
    def log_wrapping(environ, start_response):
        logging.getLogger().info('logged')
        return app(environ, start_response)
    return log_wrapping


def test_add_duplicate_middleware():
    app = WrappableApplication()
    app.middlewares.add(log_middleware, 10)
    app.middlewares.add(log_middleware, 10)
    assert list(app.middlewares) == [
        (10, log_middleware),
        (10, log_middleware),
    ]
    app.middlewares.remove(log_middleware, 10)
    assert list(app.middlewares) == [
        (10, log_middleware)
    ]


def test_add_remove_middleware():
    app = WrappableApplication()
    assert list(app.middlewares) == []
    app.middlewares.add(log_middleware, 20)
    app.middlewares.add(auth_middleware, 10)
    assert list(app.middlewares) == [
        (10, auth_middleware),
        (20, log_middleware),
    ]
    app.middlewares.add(log_middleware, 0)
    assert list(app.middlewares) == [
        (0, log_middleware),
        (10, auth_middleware),
        (20, log_middleware),
    ]

    app.middlewares.remove(auth_middleware, 10)
    assert list(app.middlewares) == [
        (0, log_middleware),
        (20, log_middleware),
    ]

    with pytest.raises(ValueError):
        app.middlewares.remove(auth_middleware, 10)

    app.middlewares.clear()
    assert list(app.middlewares) == []


def test_middleware_caller(caplog):
    app = WrappableApplication()

    @app.routes.register('/')
    def endpoint(request):
        return Response(200, body=b'Test')

    wsgi = webtest.app.TestApp(app)
    response = wsgi.get("/")
    assert response.body == b'Test'

    app.middlewares.add(auth_middleware, 10)
    response = wsgi.get("/", expect_errors=True)
    assert response.status == '401 Unauthorized'

    app.middlewares.add(log_middleware, 20)
    with caplog.at_level(logging.INFO):
        response = wsgi.get("/", expect_errors=True)
    assert response.status == '401 Unauthorized'

    with caplog.at_level(logging.INFO):
        response = wsgi.get("/", extra_environ={"REMOTE_USER": "admin"})
    assert response.body == b'Test'
    assert caplog.record_tuples == [('root', 20, 'logged')]
    caplog.clear()

    app.middlewares.remove(auth_middleware, 10)
    with caplog.at_level(logging.INFO):
        response = wsgi.get("/")
    assert response.status == '200 OK'
    assert caplog.record_tuples == [('root', 20, 'logged')]
    caplog.clear()

    app.middlewares.remove(log_middleware, 20)
    with caplog.at_level(logging.INFO):
        response = wsgi.get("/")
    assert response.status == '200 OK'
    assert caplog.record_tuples == []


src/roughrider/application/__init__.py
from typing import Optional, Type
from dataclasses import dataclass, field
from horseman.meta import Node
from horseman.types import Environ, WSGICallable, StartResponse
from roughrider.routing.components import NamedRoutes
from roughrider.application.request import Request
from roughrider.application.middleware import Middlewares


@dataclass
class Application(Node):
    routes: NamedRoutes = field(default_factory=NamedRoutes)
    utilities: dict = field(default_factory=dict)
    request_factory: Type[Request] = Request

    def resolve(self, path: str, environ: Environ) -> Optional[WSGICallable]:
        route = self.routes.match_method(path, environ['REQUEST_METHOD'])
        if route is not None:
            request = self.request_factory(self, environ, route)
            return route.endpoint(request, **route.params)


@dataclass
class WrappableApplication(Application):
    middlewares: Middlewares = None

    def __post_init__(self):
        if self.middlewares is None:
            self.middlewares = Middlewares(super().__call__)

    def __call__(self, environ: Environ, start_response: StartResponse):
        return self.middlewares(environ, start_response)


src/roughrider/application/middleware.py
import bisect
import typing as t
from functools import reduce
from horseman.types import WSGICallable, Environ, StartResponse


WSGIMiddleware = t.Callable[[WSGICallable], WSGICallable]


class Middlewares(WSGICallable):

    __slots__ = ('__call__', '_chain', 'wrapped')

    wrapped: WSGICallable
    _chain: t.List[WSGIMiddleware]

    def __init__(self, wrapped: WSGICallable):
        self.__call__ = self.wrapped = wrapped
        self._chain = []

    def update(self):
        if not self._chain:
            if self.__call__ and self.__call__ != self.wrapped:
                self.__call__ = self.wrapped
            return

        self.__call__ = reduce(
            lambda x, y: y(x),
            (m[1] for m in reversed(self._chain)),
            self.wrapped
        )

    def add(self, middleware: WSGIMiddleware, order: int = 0):
        if not self._chain:
            self._chain = [(order, middleware)]
        else:
            bisect.insort(self._chain, (order, middleware))
        self.update()

    def remove(self, middleware: WSGIMiddleware, order: int = 0):
        self._chain.remove((order, middleware))
        self.update()

    def clear(self):
        self._chain.clear()
        self.__call__ = self.wrapped

    def __iter__(self):
        return iter(self._chain)


src/roughrider/application/request.py
import typing as t
import urllib.parse
import functools
import horseman.parsers
import horseman.types
import horseman.http
import horseman.meta
from dataclasses import dataclass
from roughrider.routing.meta import Route


class Request(horseman.meta.Overhead):

    __slots__ = (
        '_content_type',
        '_cookies',
        '_data',
        '_query',
        'app',
        'environ',
        'method',
        'route',
        'script_name',
    )

    app: horseman.meta.Node
    content_type: t.Optional[horseman.http.ContentType]
    cookies: horseman.http.Cookies
    environ: horseman.types.Environ
    method: horseman.types.HTTPMethod
    query: horseman.http.Query
    route: t.Optional[Route]
    script_name: str

    _data: t.Optional[horseman.parsers.Data]

    def __init__(self,
                 app: horseman.meta.Node,
                 environ: horseman.types.Environ,
                 route: t.Optional[Route] = None):
        self._content_type = ...
        self._cookies = ...
        self._data = ...
        self._query = ...
        self.app = app
        self.environ = environ
        self.method = environ['REQUEST_METHOD'].upper()
        self.route = route
        self.script_name = urllib.parse.quote(environ['SCRIPT_NAME'])

    def extract(self) -> horseman.parsers.Data:
        if self._data is not ...:
            return self._data

        if self.content_type:
            self._data = horseman.parsers.parser(
                self.environ['wsgi.input'], self.content_type)

        return self._data

    @property
    def query(self):
        if self._query is ...:
            self._query = horseman.http.Query.from_environ(self.environ)
        return self._query

    @property
    def cookies(self):
        if self._cookies is ...:
            self._cookies = horseman.http.Cookies.from_environ(self.environ)
        return self._cookies

    @property
    def content_type(self):
        if self._content_type is ...:
            if 'CONTENT_TYPE' in self.environ:
                self._content_type = \
                    horseman.http.ContentType.from_http_header(
                        self.environ['CONTENT_TYPE'])
            else:
                self._content_type = None
        return self._content_type

    @functools.cached_property
    def application_uri(self):
        scheme = self.environ['wsgi.url_scheme']
        http_host = self.environ.get('HTTP_HOST')
        if not http_host:
            server = self.environ['SERVER_NAME']
            port = self.environ.get['SERVER_PORT']
        elif ':' in http_host:
            server, port = http_host.split(':', 1)
        else:
            server = http_host
            port = '80'

        if (scheme == 'http' and port == '80') or \
           (scheme == 'https' and port == '443'):
            return f'{scheme}://{server}{self.script_name}'
        return f'{scheme}://{server}:{port}{self.script_name}'

    def uri(self, include_query=True):
        url = self.application_uri
        path_info = urllib.parse.quote(self.environ.get('PATH_INFO', ''))
        if include_query:
            qs = urllib.parse.quote(self.environ.get('QUERY_STRING'))
            if qs:
                return f"{url}{path_info}?{qs}"
        return f"{url}{path_info}"
